// @incnt/jackson/src/components/widgets/Footer/styles.js
import {css} from 'config/css'


const styles = {
  root: {
    position: 'relative',
    boxShadow: '0 2px 3px rgba(0, 0, 0, 0.3), 0 4px 4px rgba(0, 0, 0, 0.15)',
    background: css.colors.charade,
    transform: 'translateZ(0)',
    color: 'white',
  },
  innerContainer: {
    display: 'flex',
    maxWidth: '1280px',
    flexDirection: 'column',
    boxSizing: 'border-box',
    padding: '2em 1.5em',
    maxWidth: '1440px',
    margin: '0 auto',
    [css.breakpoints.tablet]: {
      flexDirection: 'row',
      flexWrap: 'wrap',
    },
  },
  footerLeft: {
    display: 'flex',
    flexDirection: 'row',
    flexWrap: 'wrap',
    [css.breakpoints.tablet]: {
      flex: 1,
      display: 'block',
    },
  },
  footerMiddle: {
    display: 'flex',
    flexDirection: 'row',
    flexWrap: 'wrap',
    [css.breakpoints.tablet]: {
      flex: 1,
      display: 'block',
    },
  },
  footerRight: {
    [css.breakpoints.tablet]: {
      flex: '100%',
    },
    [css.breakpoints.desktop]: {
      flex: 0.3,
      display: 'block',
    },
  },
  companyInfoList: {
    display: 'flex',
    flexDirection: 'row',
    flexWrap: 'wrap',
    margin: '0 0 1.5em',
    alignItems: 'center',
    [css.breakpoints.desktop]: {
      display: 'block',
    },
  },
  footerListTitle: {
    margin: '0 0 0.5em',
    textTransform: 'uppercase',
    letterSpacing: '0.75px',
    flex: '100%',
    [css.breakpoints.desktop]: {
      flex: '1',
    },
  },
  footerListHeader: {
    fontSize: '1em',
  },
  link: {
    flex: '100%',
    whiteSpace: 'nowrap',
    fontWeight: 300,
    lineHeight: 2,
    [css.breakpoints.desktop]: {
      flex: 1,
    },
  },
  logoWrapper: {
    display: 'flex',
    margin: '1em 0 0',
  },

}

export {styles}
