// @incnt/jackson/webpack.config.js
const path = require('path')
const nodeExternals = require('webpack-node-externals')

module.exports = {
  entry: {
    index: './index.js',
  },
  output: {
    path: path.resolve(__dirname, 'dist'),
    filename: '[name].js',
    sourceMapFilename: '[name].js.map',
    library: '',
    libraryTarget: 'commonjs',
  },
  target: 'node',
  externals: [nodeExternals()],
  resolve: {
    modules: ['src', 'node_modules'],
  },
  devtool: 'source-map',
  module: {
    rules: [
      {
        test: /\.js$/,
        exclude: /node_modules/,
        use: {
          loader: 'babel-loader',
        },
      },
      {
        test: /\.(scss|sass)$/,
        loaders: ['style-loader', 'css-loader', 'sass-loader'],
      },
      {
        test: /\.html$/,
        loaders: ['html-loader'],
      },
      {
        test: /\.*$/,
        loaders: ['file-loader'],
        include: path.resolve(__dirname, 'src/assets'),
      },
      {
        test: /\.css$/,
        loaders: ['style-loader', 'css-loader'],
      },
    ],
  },
}
