// @incnt/jackson/src/components/forms/inputs/CountryInput/index.js
import React       from 'react'
import injectSheet from 'react-jss'
import {countries} from 'config/countries'
import _           from 'lodash'
import {css}       from 'config/css'

const CountryInput = injectSheet({
  root: {
    position: 'relative',
  },
  input: {
    display: 'flex',
    fontSize: '1.5em',
    alignItems: 'center',
    marginBottom: '0.5rem',
    justifyContent: 'center',
  },
  selectedFlag: {
    marginRight: '0.5rem',
    display: 'inline-block',
  },
  results: {},
  mobile: {
    [css.breakpoints.desktop]: {
      display: 'none',
    },
  },
  desktop: {
    [css.breakpoints.maxDesktop]: {
      display: 'none',
    },
  },
  countyInput:{
    width: '30%',
    textAlign: 'center'
  }
})(
  class CountryInput extends React.Component {
    constructor(props) {
      super(props)
      this.state = {
        filter: '',
        country: null,
      }
    }

    selectCountry(country) {
      this.setState({country, filter: country.name})
      if (typeof this.props.onChange === 'function') this.props.onChange(country)
    }

    render() {
      const {classes} = this.props
      const {country, filter} = this.state
      let show = true
      if (!filter || (filter && country && country.name === filter)) show = false
      return (
        <div className={classes.root}>
          <div className={classes.input}>
            {country &&
            <Flag className={`${classes.selectedFlag} ${classes.desktop}`}
                  size="32"
                  code={country.al.toLowerCase()}/>}
            <Input
              type='text'
              placeholder="Country"
              value={filter}
              onChange={e => this.setState({filter: e.target.value})}
            />
          </div>
          {show && <ul className={classes.results}>
            <Results clickHandler={country => this.selectCountry(country)} filter={this.state.filter}/>
          </ul>}
        </div>
      )
    }
  },
)

const Input = injectSheet({
  countyInput:{
    width: '30%',
    textAlign: 'center'
  }
})(({classes, ...rest}) => (
  <input
    className={classes.countyInput}
    type="text"
    {...rest}
  />
))

const Results = injectSheet({
  result: {
    fontSize: '0.9em',
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    borderBottom: '1px solid rgba(0, 0, 0, 0.1)',
    padding: '0.25rem 0.75rem',
    '&:last-child': {
      borderBottom: 'none',
    },
  },
  flag: {
    marginRight: '0.75rem',
  },
})(({classes, clickHandler, filter, ...rest}) => {
  const filteredCountries = filterCountries(filter, countries)
  return (
    <React.Fragment>
      {_.map(filteredCountries, country => (
        <li className={classes.result}
            key={country.name + country.cc + country.al}
            onClick={() => clickHandler(country)}
            {...rest}>
          <Flag className={classes.flag} code={country.al.toLowerCase()}/>
          <span>{country.name}</span>
        </li>
      ))}
    </React.Fragment>
  )
})


export {CountryInput}

// al:"SC"
// cc:["248"]
// name:"Seychelles"

function filterCountries(needle, haystack) {
  if (!needle) return haystack
  return haystack.filter(hay => hay.name.toLowerCase().includes(needle.toLowerCase()))
}

const Flag = ({code, size = '24', ...rest}) => (
  <img src={`https://www.countryflags.io/${code}/flat/${size}.png`} alt={code} {...rest}/>
)
