// @incnt/jackson/src/components/layout/Column.js
import React       from 'react'
import injectSheet from 'react-jss'
import {css}       from 'config/css'
import {objstr}    from 'config/objectString'

const styles = {
  root: {
    flexBasis: '100%',
    flexGrow: 0,
    flexShrink: 0,
    margin: '0 0 0.5rem',
    padding: '0 0.25rem',
    minWidth: 0,
    [css.breakpoints.tablet]: {
      flexBasis: 1,
      flexGrow: 1,
    },
  },
  col: {
    flexGrow: 0,
    [css.breakpoints.tablet]: {
      flexGrow: 0,
    },
  },
  ...generateColumns('xs', null),
  ...generateColumns('sm', 480),
  ...generateColumns('md', 768),
  ...generateColumns('lg', 1024),
  ...generateColumns('xl', 1280),
}

function generateColumns(prefix, width) {
  const classes = {}
  for (let i = 1; i <= 12; i++) {
    classes[`${prefix}${i}`] = {
      flexBasis: `${(i / 12) * 100}%`,
    }
  }
  return (width) ? {[`@media(min-width: ${width}px)`]: {...classes}} : classes
}

const Column = injectSheet(styles)(({className, classes, children, xs, sm, md, lg, xl, ...rest}) => {
  const classMap = {
    [className]: true,
    [classes.root]: true,
    [classes.col]: (xs || sm || md || lg || xl),
    [classes[`xs${xs}`]]: !!xs,
    [classes[`sm${sm}`]]: !!sm,
    [classes[`md${md}`]]: !!md,
    [classes[`lg${lg}`]]: !!lg,
    [classes[`xl${xl}`]]: !!xl,
  }

  return (
    <div className={objstr(classMap)} {...rest}>
      {children}
    </div>
  )
})

export {Column}
