// @incnt/jackson/src/components/forms/inputs/SelectInput/styles.js
import {css} from 'config/css'

const styles = {
  root: {
    fontWeight: 300,
    position: 'relative',
  },
  search: {},
  options: {
    marginTop: '0.25em',
    background: css.colors.white,
    color: css.colors.charade,
    maxHeight: '300px',
    overflowY: 'auto',
    boxShadow: css.shadows.low,
    borderRadius: '3px',
    display: 'none',
    [css.breakpoints.tablet]: {
      position: 'absolute',
      zIndex: 1,
      top: '100%',
      left: 0,
      right: 0,
    },
  },
  optionsStatic: {
    position: 'static !important',
  },
  optionsActive: {
    display: 'block',
  },
}
const searchStyles = {
  root: {
    position: 'relative',
  },
  input: {
    width: '100%',
    padding: '0 2.75em 0 1em',
    lineHeight: 3,
    boxShadow: css.shadows.low,
    borderRadius: '3px',
    background: css.colors.white,
    transition: css.transitions.fast,
    '&:focus': {
      background: css.colors.alabaster,
    },
    '&:hover': {
      background: css.colors.alabaster,
    },
  },
  actions: {
    position: 'absolute',
    right: 0,
    top: 0,
    height: '100%',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },
  action: {
    width: '2.75em',
    position: 'relative',
    display: 'flex',
    height: '100%',
    justifyContent: 'center',
    alignItems: 'center',
    flex: 1,
    textAlign: 'center',
    background: css.colors.white,
    color: css.colors.black,
    transition: css.transitions.fast,
    borderLeft: `1px solid ${css.colors.alto}`,
    cursor: 'pointer',
    '&:hover': {
      background: css.colors.seafarer,
      color: css.colors.white,
    },
    '&:before, &:after': {
      position: 'absolute',
      display: 'block',
      content: '""',
      left: '50%',
      top: '50%',
      transition: css.transitions.slow,
    },
  },
  deleteAction: {
    '&:before, &:after': {
      width: '14px',
      height: '2px',
      boxShadow: `inset 0 0 0 10px ${css.colors.charade}`,
      borderRadius: '2px',
    },
    '&:hover:before, &:hover:after': {
      boxShadow: `inset 0 0 0 10px ${css.colors.white}`,
    },
    '&:before': {
      transform: 'translate(-50%, -50%) rotate(-45deg)',
    },
    '&:after': {
      transform: 'translate(-50%, -50%) rotate(45deg)',
    }
  },
  toggleAction: {
    borderRadius: '0 3px 3px 0',
  },
  close: {
    '&:before, &:after': {
      width: '10px',
      height: '2px',
      boxShadow: `inset 0 0 0 10px ${css.colors.charade}`,
      borderRadius: '2px',
    },
    '&:hover:before, &:hover:after': {
      boxShadow: `inset 0 0 0 10px ${css.colors.white}`,
    },
    '&:before': {
      transform: 'translate(-80%, -50%) rotate(-225deg)',
    },
    '&:after': {
      transform: 'translate(-20%, -50%) rotate(225deg)',
    },
  },
  open: {
    '&:before, &:after': {
      width: '10px',
      height: '2px',
      boxShadow: `inset 0 0 0 10px ${css.colors.charade}`,
      borderRadius: '2px',
    },
    '&:hover:before, &:hover:after': {
      boxShadow: `inset 0 0 0 10px ${css.colors.white}`,
    },
    '&:before': {
      transform: 'translate(-80%, -50%) rotate(45deg)',
    },
    '&:after': {
      transform: 'translate(-20%, -50%) rotate(-45deg)',
    },
  },
}
const optionsStyles = {
  root: {
    padding: '0.5em 0',
  },
}
const optionStyles = {
  root: {
    cursor: 'pointer',
    padding: '1em 1.25em',
    borderBottom: `1px solid ${css.colors.alabaster}`,
    '&:hover': {
      background: css.colors.alabaster,
    },
    '&:last-child': {
      borderBottom: 'none',
    },
  },
  active: {},
  inactive: {
    display: 'none',
  },
}

export {styles, searchStyles, optionsStyles, optionStyles}
