// @incnt/jackson/src/components/content/Dropdown.js
import React       from 'react'
import injectSheet from 'react-jss'
import {objstr}    from 'config/objectString'

const Dropdown = injectSheet({
  dropdown: {},
})(class Dropdown extends React.Component {
  constructor(props) {
    super(props)
    this.state = {}
  }

  render() {
    const {classes, Parent, children} = this.props
    return (
      <div className={classes.dropdown}>
        <Toggle open={this.state.open} Component={Parent} onClick={() => this.setState({open: !this.state.open})}/>
        <Children open={this.state.open}>{children}</Children>
      </div>
    )
  }
})

const Toggle = injectSheet({
  root: {
    cursor: 'pointer',
    display: 'inline-flex',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  // caret: {
  //   width: 0,
  //   height: 0,
  //   margin: '0 0 0 0.5em',
  //   borderLeft: '5px solid transparent',
  //   borderRight: '5px solid transparent',
  // },
  // active: {
  //   borderBottom: '8px solid black',
  // },
  // inactive: {
  //   borderTop: '8px solid black',
  // },
})(({classes, Component, onClick, open, ...rest}) => (
  <div className={classes.root} onClick={onClick}>
    <Component open={open} {...rest}/>
    <div className={objstr({
      [classes.caret]: true,
      [classes.active]: !!open,
      [classes.inactive]: !open,
    })}/>
  </div>
))

const Children = injectSheet({
  root: {
    position: 'absolute',
    width: '100%',
  },
  active: {
    display: 'block',
  },
  inactive: {
    display: 'none',
  },
})(({classes, children, open, ...rest}) => (
  <div className={objstr({
    [classes.root]: true,
    [classes.active]: !!open,
    [classes.inactive]: !open,
  })} {...rest}>
    {children}
  </div>
))

export {Dropdown}
