// @incnt/jackson/src/components/forms/inputs/DateInput/styles.js
import {css} from 'config/css'
import color from 'color'

const coloredBackground = css.colors.seafarer
const coloredFontColor = css.colors.white
const defaultBackground = css.colors.white
const defaultFontColor = css.colors.black
const caretBackground = css.colors.alabaster
const caretFontColor = css.colors.black

const styles = {
  root: {
    width: '100%',
    background: defaultBackground,
    color: defaultFontColor,
    fontWeight: '300',
    filter: `drop-shadow(${css.shadows.low})`,
    borderRadius: '0 0 3px 3px',
    transition: css.transitions.fast,
    position: 'relative',
  },
  active: {
    opacity: 1,
    height: 'auto',
  },
  inactive: {
    opacity: 0,
    height: 0,
    overflow: 'hidden',
  },
  header: {},
  calendar: {
    background: css.colors.white,
  },
  sliders: {
    display: 'flex',
    justifyContent: 'space-between',
    background: color(coloredBackground).hex(),
    color: coloredFontColor,
    padding: '0 0.5em',
  },
  yearSlider: {
    display: 'block',
    width: '100%',
  },
  monthSlider: {
    display: 'block',
    width: '100%',
  },
  dayLabels: {},
  daySelector: {},
  actions: {},
  static: {
    [css.breakpoints.tablet]: {
      zIndex: 1,
      width: '100%',
      position: 'relative',
    },
  },
}
const headerStyles = {
  root: {
    position: 'relative',
    padding: '0.8rem 1rem',
    cursor: 'pointer',
    transition: css.transitions.fast,
    '&:before, &:after': {
      position: 'absolute',
      display: 'block',
      content: '""',
      right: '1rem',
      top: '50%',
      transition: css.transitions.fast,
      borderRadius: '2px',
      height: '2px',
    },
  },
  active: {
    borderRadius: '3px 3px 0 0',
    background: color(coloredBackground).darken(0.2).hex(),
    color: coloredFontColor,
    fontSize: '1.25em',
    '&:hover': {
      background: color(coloredBackground).darken(0.4).hex(),
    },
    '&:before, &:after': {
      width: '14px',
      boxShadow: `inset 0 0 0 10px ${css.colors.white}`,
    },
    '&:before': {
      transform: 'translate(-50%, -50%) rotate(-45deg)',
    },
    '&:after': {
      transform: 'translate(-50%, -50%) rotate(45deg)',
    },
  },
  inactive: {
    background: css.colors.white,
    borderRadius: '3px',
    '&:hover': {
      background: css.colors.porcelain,
    },
    '&:before, &:after': {
      width: '10px',
      boxShadow: `inset 0 0 0 10px ${css.colors.charade}`,
    },
    '&:before': {
      transform: 'translate(-80%, -50%) rotate(45deg)',
    },
    '&:after': {
      transform: 'translate(-20%, -50%) rotate(-45deg)',
    },
  },
  label: {
    display: 'inline-block',
    marginRight: '0.25em',
  },
  activeLabel: {
    display: 'none',
  },
  inactiveLabel: {},
  date: {
    fontWeight: 400,
  },
}
const yearSliderStyles = {
  root: {
    width: '100%',
    minWidth: '80px',
    maxWidth: '130px',
    fontSize: '0.85em',
    padding: '0.75em 0',
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  caret: {
    cursor: 'pointer',
    background: caretBackground,
    color: caretFontColor,
    padding: '0.5em',
    margin: '0 0.25em',
    lineHeight: '0.5em',
    borderRadius: '100%',
    transition: css.transitions.fast,
    '&:hover': {
      background: caretFontColor,
      color: caretBackground,
    },
  },
  leftCaret: {},
  rightCaret: {},
  label: {},
}
const monthSliderStyles = {
  root: {
    width: '100%',
    minWidth: '120px',
    maxWidth: '130px',
    marginLeft: 'auto',
    fontSize: '0.85em',
    padding: '0.75em 0',
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  caret: {
    cursor: 'pointer',
    background: caretBackground,
    color: caretFontColor,
    padding: '0.5em',
    margin: '0 0.25em',
    lineHeight: '0.5em',
    borderRadius: '100%',
    transition: css.transitions.fast,
    '&:hover': {
      background: caretFontColor,
      color: caretBackground,
    },
  },
  rightCaret: {},
  leftCaret: {},
  label: {},
}
const dayLabelStyles = {
  root: {
    display: 'flex',
    flexWrap: 'wrap',
    background: css.colors.porcelain,
  },
  day: {
    flex: `${100 / 7}% 0 0`,
    textAlign: 'center',
    padding: '1em 0',
    fontSize: '0.75em',
  },
}
const daySelectorStyles = {
  root: {
    display: 'flex',
    flexWrap: 'wrap',
  },
}
const dayStyles = {
  root: {
    flex: `${100 / 7}% 0 0`,
    textAlign: 'center',
    cursor: 'pointer',
  },
  selected: {},
  label: {
    fontSize: '0.75em',
    width: '3em',
    height: '3em',
    lineHeight: '3em',
    padding: '0',
    margin: '0 auto',
    borderRadius: '100%',
    transition: css.transitions.fast,
    '&:hover': {
      background: caretBackground,
      color: caretFontColor,
    },
  },
  selectedLabel: {
    background: coloredBackground,
    color: coloredFontColor,
    '&:hover': {
      background: coloredBackground,
      color: coloredFontColor,
    },
  },
}
const actionStyles = {
  root: {
    display: 'flex',
    justifyContent: 'flex-end',
  },
  button: {
    color: css.colors.seafarer,
    textTransform: 'uppercase',
    margin: '1em 0 0',
    padding: '0.5em 1em',
    cursor: 'pointer',
    '&:hover': {
      color: css.colors.charade,
    },
  },
}

export {
  styles,
  headerStyles,
  yearSliderStyles,
  monthSliderStyles,
  dayLabelStyles,
  daySelectorStyles,
  dayStyles,
  actionStyles,
}
