// @incnt/jackson/src/components/content/Card/index.js
import React        from 'react'
import injectSheet  from 'react-jss'
import {objstr}     from 'config/objectString'
import {CardHeader} from 'components/content/Card/CardHeader'
import {CardBody}   from 'components/content/Card/CardBody'
import {CardFooter} from 'components/content/Card/CardFooter'
import {
  card,
  cardDarkTheme,
  cardLightTheme,
}                   from 'components/content/cardStyles'

const Card = injectSheet({
  card: {
    ...card,
    margin: '0 0 1rem',
    width: '100%',
  },
  light: {...cardLightTheme},
  dark: {...cardDarkTheme},
})(({className, classes, children, theme, ...rest}) => {
  const c = objstr({
    [className]: !!className,
    [classes.card]: true,
    [classes.dark]: theme === 'dark',
    [classes.light]: !theme || theme === 'light',
  })
  return (
    <div className={c}>
      {React.Children.map(children, child => {
        if (React.isValidElement(child)) {
          return React.cloneElement(child, {theme, ...rest})
        } else {
          return child
        }
      })}
    </div>
  )
})

export {Card, CardHeader, CardBody, CardFooter}
