// @incnt/jackson/src/components/forms/inputs/PhoneInput/styles.js
import {css} from 'config/css'

const styles = {
  root: {
    display: 'flex',
    fontWeight: 300,
    position: 'relative',
    borderBottom: '1px solid rgba(0, 0, 0, 0.1)'
  },
  countryInput: {
    display: 'flex',
    flex: '1 0 100px',
    minWidth: 0,
  },
  countryCode: {
    flex: '0 0 0',
    minWidth: '0',
    visibility: 'hidden',
    opacity: '0',
    padding: '0',
    lineHeight: '3em',
    // boxShadow: css.shadows.low,
    // borderRadius: '3px 0 0 3px',,
    background: css.colors.white,
    transition: css.transitions.fast,
    transform: 'translateZ(0)',
    cursor: 'pointer',
    '&:focus': {
      background: css.colors.alabaster,
    },
    '&:hover': {
      background: css.colors.alabaster,
    },
  },
  countryCodeActive: {
    padding: '0 1em',
    flex: '0 1 100px',
    visibility: 'visible',
    opacity: '1',
  },
  countrySelect: {
    flex: '0 0 0',
    minWidth: '0',
    visibility: 'hidden',
    opacity: '0',
    transition: css.transitions.fast,
    transform: 'translateZ(0)',
  },
  countrySelectActive: {
    flex: 1,
    visibility: 'visible',
    opacity: 1,
    transition: css.transitions.slow,
  },
  numberInput: {
    flex: '0 0 0',
    visibility: 'hidden',
    overflow: 'hidden',
    opacity: '0',
    transition: css.transitions.fast,
    transform: 'translateZ(0)',
  },
  numberInputActive: {
    flex: '1 1 100%',
    visibility: 'visible',
    overflow: 'initial',
    opacity: 1,
    transition: css.transitions.slow,
  },
}

const countryCodeStyles = {
  root: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
  image: {
    width: '1.5em',
    height: '0.75em',
  },
  callingCode: {
    display: 'inline-block',
    padding: '0 0 0 0.25em',
  },
  name: {
    display: 'none',
  },
}

const countrySelectStyles = {
  root: {},
}

const numberInputStyles = {
  root: {},
  input: {
    width: '100%',
    padding: '0 1em',
    lineHeight: '3em',
    // boxShadow: css.shadows.low,
    // borderRadius: '0 3px 3px 0',
    background: css.colors.white,
    transition: css.transitions.fast,
    transform: 'translateZ(0)',
    '&:focus': {
      background: css.colors.alabaster,
    },
    '&:hover': {
      background: css.colors.alabaster,
    },
  },
}

const countryLabelStyles = {
  root: {
    display: 'flex',
  },
  image: {
    width: '1.5em',
    height: '1.5em',
    margin: '0 1em 0 0',
  },
  name: {
    fontWeight: 700,
    margin: '0 1em 0 0',
    whiteSpace: 'nowrap',
  },
  code: {},
}

export {styles, countrySelectStyles, countryCodeStyles, numberInputStyles, countryLabelStyles}
