// @incnt/jackson/src/components/widgets/Menus/MenuItem.js
import React       from 'react'
import injectSheet from 'react-jss'
import {css}       from 'config/css'
import color       from 'color'
import {objstr}    from 'config/objectString'

function isTextLink(props) {
  return (props.to || props.href)
}

// ({href, to}) => val >> this is to catch text links and make them look pretty
const menuItemStyles = {
  dark: {
    borderBottom: `1px solid ${color(css.colors.charade).lighten(0.5).hex()}`,
  },
  light: {
    borderBottom: `1px solid ${color(css.colors.white).darken(0.1).hex()}`,
  },
  menuItem: {
    flex: '100%',
    whiteSpace: 'nowrap',
    cursor: 'pointer',
    '&:last-child': {
      borderBottom: 'none',
    },
    [css.breakpoints.desktop]: {
      flex: 1,
      margin: '0 0.5rem',
      borderBottom: 'none',
    },
  },
  notTextLink: {
    // margin: '0 0.5rem 0.5rem',
    // [css.breakpoints.tablet]: {
    //   margin: '0 1rem 1rem',
    // },
  },
  textLink: {
    padding: '0.5rem',
    lineHeight: 2,
    [css.breakpoints.tablet]: {
      padding: '1rem',
    },
    [css.breakpoints.desktop]: {
      padding: 0,
    },
  },
}

const MenuItem = injectSheet(menuItemStyles)(
  ({classes, className, children, theme, active, Component = 'a', loggedIn, withLogin, to, href, flex, mobileFlex, closeNav,...rest}) => {
    // with login shouldn't exist, or if it does should be something available in all our components as sensible
    if (!withLogin || (withLogin && withLogin(loggedIn))) {
      const classMap = {
        [className]: !!className,
        [classes.menuItem]: true,
        [classes.dark]: (theme === 'dark'),
        [classes.light]: (!theme || theme === 'light'),
        [classes.textLink]: (to || href),
        [classes.notTextLink]: !(to || href),
      }
      return (
        <Component className={objstr(classMap)} onClick={()=>closeNav()} to={to} href={href} {...rest}>
          {children}
        </Component>
      )
    }
    return null
    // if (!withLogin || typeof withLogin === 'function' && withLogin(loggedIn)) {
    //   return (
    //     <Component className={`${classes.root} ${className}`} {...rest}>
    //       {`${loggedIn}`}
    //     </Component>
    //   )
    // }
    // return null
  },
)

export {MenuItem}
