// @incnt/jackson/src/components/widgets/Menus/Menu.js
import React       from 'react'
import injectSheet from 'react-jss'
import {css}       from 'config/css'
import {objstr}    from 'config/objectString'

const styles = {
  dark: {
    background: css.colors.charade,
    color: css.colors.white,
  },
  light: {
    background: css.colors.white,
    color: css.colors.charade,
  },
  menu: {
    display: 'flex',
    width: '100%',
    flexWrap: 'wrap',
    margin: '1rem 0',
    [css.breakpoints.desktop]: {
      width: 'auto',
      alignItems: 'center',
      margin: '0 0.5rem',
    },
  },
}

const Menu = injectSheet(styles)(({className, classes, active = true, children, theme, ...rest}) => (
  <nav className={objstr({
    [classes.menu]: true,
    [className]: !!className,
    [classes.dark]: (theme === 'dark'),
    [classes.light]: (!theme || theme === 'light'),
  })}>
    {React.Children.map(children, child => {
      if (typeof child === 'string') return child
      if (React.isValidElement(child)) return React.cloneElement(child, {theme, ...rest})
      return null
    })}
  </nav>
))

export {Menu}
