// @incnt/jackson/src/components/forms/buttons/Button.js
import React       from 'react'
import injectSheet from 'react-jss'
import color       from 'color'
import {css}       from 'config/css'

function getClassNames(object) {
  const classNames = []
  for (const name in object) {
    const condition = object[name]
    if ((typeof condition === 'function') ? condition() : condition) {
      classNames.push(name)
    }
  }
  return classNames.length ? classNames.join(' ') : null
}

const styles = {
  button: {
    fontSize: '1em',
    padding: '0.5rem 1rem',
    borderRadius: css.boxModel.roundCorners,
    background: css.colors.seafarer,
    color: css.colors.white,
    boxShadow: css.shadows.low,
    cursor: 'pointer',
    transition: css.transitions.fast,
    '&:hover': {
      background: color(css.colors.seafarer).lighten(0.1).hex(),
      color: color(css.colors.white).lighten(0.1).hex(),
      boxShadow: css.shadows.lowHover,
    },
  },
  largeButton: {
    padding: '1rem 1.5rem',
    fontSize: '1.5em',
  },
  invertedButton: {
    background: css.colors.white,
    color: css.colors.seafarer,
    '&:hover': {
      background: color(css.colors.white).lighten(0.1).hex(),
      color: color(css.colors.seafarer).lighten(0.1).hex(),
    },
  },
}

const Button = injectSheet(styles)(({className, classes, onClick, label, children, large, inverted, ...rest}) => {
  const classMap = {
    [className]: true,
    [classes.button]: true,
    [classes.largeButton]: () => !!large,
    [classes.invertedButton]: () => !!inverted,
  }

  return (
    <button className={getClassNames(classMap)} onClick={onClick} {...rest}>
      {label || children}
    </button>
  )
})

export {Button}
